File: client_java/NetworkManager.java
/**
 * Network communication manager
 * Handles TCP connection and communication protocol
 */
import java.io.*;
import java.net.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.atomic.AtomicBoolean;

public class NetworkManager {
    private Socket socket;
    private BufferedReader in;
    private PrintWriter out;
    private AtomicBoolean connected = new AtomicBoolean(false);
    private AtomicBoolean authenticated = new AtomicBoolean(false);
    private AtomicBoolean isAdmin = new AtomicBoolean(false);
    private String username = "";
    private Thread receiveThread;
    
    // Callbacks for network events
    public interface NetworkEventListener {
        void onConnected();
        void onDisconnected();
        void onAuthenticationSuccess();
        void onAuthenticationFailed();
        void onDataReceived(String data);
        void onError(String error);
    }
    
    private NetworkEventListener listener;
    
    public NetworkManager(NetworkEventListener listener) {
        this.listener = listener;
    }
    
    public boolean connect(String host, int port) {
        try {
            socket = new Socket(host, port);
            in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            out = new PrintWriter(socket.getOutputStream(), true);
            
            connected.set(true);
            username = "";
            authenticated.set(false);
            isAdmin.set(false);
            
            // Start thread to receive messages
            receiveThread = new Thread(this::receiveMessages);
            receiveThread.setDaemon(true);
            receiveThread.start();
            
            if (listener != null) {
                listener.onConnected();
            }
            
            return true;
        } catch (IOException e) {
            if (listener != null) {
                listener.onError("Error connecting: " + e.getMessage());
            }
            return false;
        }
    }
    
    public void disconnect() {
        if (connected.get()) {
            try {
                sendCommand("DISCONNECT:");
                connected.set(false);
                authenticated.set(false);
                isAdmin.set(false);
                username = "";
                
                if (socket != null) {
                    socket.close();
                }
                
                if (listener != null) {
                    listener.onDisconnected();
                }
            } catch (IOException e) {
                if (listener != null) {
                    listener.onError("Error desconectando: " + e.getMessage());
                }
            }
        }
    }
    
    public void authenticate(String username, String password) {
        if (!connected.get()) {
            if (listener != null) {
                listener.onError("No connection to server");
            }
            return;
        }
        
        this.username = username;
        sendCommand("AUTH: " + username + " " + password);
    }
    
    public void requestData() {
        if (!connected.get()) {
            if (listener != null) {
                listener.onError("No connection to server");
            }
            return;
        }
        sendCommand("GET_DATA:");
    }
    
    public void sendVehicleCommand(String command) {
        if (!connected.get()) {
            if (listener != null) {
                listener.onError("No connection to server");
            }
            return;
        }
        
        if (!isAdmin.get()) {
            if (listener != null) {
                listener.onError("Only administrators can send commands");
            }
            return;
        }
        
        sendCommand("SEND_CMD: " + command);
    }
    
    public void requestUsersList() {
        if (!connected.get()) {
            if (listener != null) {
                listener.onError("No connection to server");
            }
            return;
        }
        
        if (!isAdmin.get()) {
            if (listener != null) {
                listener.onError("Only administrators can view users");
            }
            return;
        }
        
        sendCommand("LIST_USERS:");
    }
    
    private void sendCommand(String command) {
        if (out != null) {
            String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            String message = command + "\r\nUSER: " + username + "\r\nTIMESTAMP: " + timestamp + "\r\n\r\n";
            out.println(message);
        }
    }
    
    private void receiveMessages() {
        try {
            String line;
            StringBuilder message = new StringBuilder();
            
            while (connected.get() && (line = in.readLine()) != null) {
                if (line.trim().isEmpty()) {
                    // Fin del mensaje
                    processServerMessage(message.toString());
                    message = new StringBuilder();
                } else {
                    message.append(line).append("\n");
                }
            }
        } catch (IOException e) {
            if (connected.get() && listener != null) {
                listener.onError("Error recibiendo mensajes: " + e.getMessage());
                disconnect();
            }
        }
    }
    
    private void processServerMessage(String message) {
        if (listener != null) {
            listener.onDataReceived(message);
        }
        
        if (message.startsWith("AUTH_SUCCESS")) {
            authenticated.set(true);
            isAdmin.set(true);
            if (listener != null) {
                listener.onAuthenticationSuccess();
            }
        } else if (message.startsWith("AUTH_FAILED")) {
            authenticated.set(false);
            isAdmin.set(false);
            if (listener != null) {
                listener.onAuthenticationFailed();
            }
        }
    }
    
    // Getters para estado
    public boolean isConnected() { return connected.get(); }
    public boolean isAuthenticated() { return authenticated.get(); }
    public boolean isAdmin() { return isAdmin.get(); }
    public String getUsername() { return username; }
}

File: client_python/network_manager.py
#!/usr/bin/env python3
"""
Gestor de comunicación de red
Maneja la conexión TCP y el protocolo de comunicación
"""

import socket
import threading
from datetime import datetime
from typing import Callable, Optional

class NetworkManager:
    def __init__(self):
        self.host = 'localhost'
        self.port = 8080
        self.socket = None
        self.connected = False
        self.authenticated = False
        self.is_admin = False
        self.username = ""
        self.running = True
        
        # Callbacks para eventos
        self.on_connected: Optional[Callable] = None
        self.on_disconnected: Optional[Callable] = None
        self.on_authentication_success: Optional[Callable] = None
        self.on_authentication_failed: Optional[Callable] = None
        self.on_data_received: Optional[Callable[[str], None]] = None
        self.on_error: Optional[Callable[[str], None]] = None
        
        # Hilo de recepción
        self.receive_thread = None
    
    def connect(self, host: str = 'localhost', port: int = 8080) -> bool:
        """Conectar al servidor"""
        try:
            self.host = host
            self.port = port
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.settimeout(10)  # Timeout de 10 segundos
            self.socket.connect((host, port))
            self.connected = True
            self.authenticated = False
            self.is_admin = False
            self.username = ""
            
            # Iniciar hilo para recibir mensajes
            self.receive_thread = threading.Thread(target=self._receive_messages)
            self.receive_thread.daemon = True
            self.receive_thread.start()
            
            if self.on_connected:
                self.on_connected()
            
            return True
        except Exception as e:
            if self.on_error:
                self.on_error(f"Error conectando: {str(e)}")
            return False
    
    def disconnect(self):
        """Desconectar del servidor"""
        try:
            if self.connected:
                self._send_command("DISCONNECT:")
                self.connected = False
                self.authenticated = False
                self.is_admin = False
                self.running = False
                self.socket.close()
                if self.on_disconnected:
                    self.on_disconnected()
        except Exception as e:
            if self.on_error:
                self.on_error(f"Error desconectando: {str(e)}")
    
    def authenticate(self, username: str, password: str) -> bool:
        """Autenticar como administrador"""
        if not self.connected:
            if self.on_error:
                self.on_error("No hay conexión con el servidor")
            return False
        
        try:
            self.username = username
            command = f"AUTH: {username} {password}"
            self._send_command(command)
            return True
        except Exception as e:
            if self.on_error:
                self.on_error(f"Error en autenticación: {str(e)}")
            return False
    
    def request_data(self) -> bool:
        """Solicitar datos de telemetría"""
        if not self.connected:
            if self.on_error:
                self.on_error("No hay conexión con el servidor")
            return False
        return self._send_command("GET_DATA:")
    
    def send_vehicle_command(self, cmd: str) -> bool:
        """Enviar comando de control del vehículo"""
        if not self.connected:
            if self.on_error:
                self.on_error("No hay conexión con el servidor")
            return False
        
        if not self.is_admin:
            if self.on_error:
                self.on_error("Solo administradores pueden enviar comandos")
            return False
        
        valid_commands = ["SPEED_UP", "SLOW_DOWN", "TURN_LEFT", "TURN_RIGHT"]
        if cmd not in valid_commands:
            if self.on_error:
                self.on_error(f"Comando no válido: {cmd}")
            return False
        
        return self._send_command(f"SEND_CMD: {cmd}")
    
    def request_users_list(self) -> bool:
        """Solicitar lista de usuarios conectados"""
        if not self.connected:
            if self.on_error:
                self.on_error("No hay conexión con el servidor")
            return False
        
        if not self.is_admin:
            if self.on_error:
                self.on_error("Solo administradores pueden ver usuarios")
            return False
        
        return self._send_command("LIST_USERS:")
    
    def _send_command(self, command: str) -> bool:
        """Enviar comando al servidor"""
        if not self.connected or not self.socket:
            return False
        
        try:
            # Formatear mensaje según protocolo
            timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            message = f"{command}\r\nUSER: {self.username}\r\nTIMESTAMP: {timestamp}\r\n\r\n"
            
            self.socket.send(message.encode('utf-8'))
            return True
        except Exception as e:
            if self.on_error:
                self.on_error(f"Error enviando comando: {str(e)}")
            return False
    
    def _receive_messages(self):
        """Hilo para recibir mensajes del servidor"""
        while self.connected and self.running:
            try:
                data = self.socket.recv(1024)
                if not data:
                    if self.on_error:
                        self.on_error("Servidor cerró la conexión")
                    break
                
                message = data.decode('utf-8').strip()
                self._process_server_message(message)
                
            except socket.timeout:
                continue
            except Exception as e:
                if self.connected and self.on_error:
                    self.on_error(f"Error recibiendo mensaje: {str(e)}")
                break
    
    def _process_server_message(self, message: str):
        """Procesar mensaje recibido del servidor"""
        try:
            if self.on_data_received:
                self.on_data_received(message)
            
            if message.startswith("AUTH_SUCCESS"):
                self.authenticated = True
                self.is_admin = True
                if self.on_authentication_success:
                    self.on_authentication_success()
                    
            elif message.startswith("AUTH_FAILED"):
                self.authenticated = False
                self.is_admin = False
                if self.on_authentication_failed:
                    self.on_authentication_failed()
                    
        except Exception as e:
            if self.on_error:
                self.on_error(f"Error procesando mensaje: {str(e)}")
    
    # Getters para estado
    def is_connected(self) -> bool:
        return self.connected
    
    def is_authenticated(self) -> bool:
        return self.authenticated
    
    def is_admin_user(self) -> bool:
        return self.is_admin
    
    def get_username(self) -> str:
        return self.username
